library/bignum_core.c

```diff
@@ -353,6 +353,46 @@ void mbedtls_mpi_core_shift_r(mbedtls_mpi_uint *X, size_t limbs,
     }
 }
 
+void mbedtls_mpi_core_shift_l(mbedtls_mpi_uint *X, size_t limbs,
+                              size_t count)
+{
+    size_t i, v0, v1;
+    mbedtls_mpi_uint r0 = 0, r1;
+
+    v0 = count / (biL);
+    v1 = count & (biL - 1);
+
+    if (v0 > limbs || (v0 == limbs && v1 > 0)) {
```

Comment:
I think that method was intended to catch incorrect input cases, where count is (N * bil) - 1:

For example for limbs = 1, count = 127, bil = 64:

~~~~~
v0 = 127/ 64                       ==> 1
v1 = 127& 63                       ==> 63
limbs = 1
v0 > limbs                         ==> True
v0 == limbs ==> 1 == 1             ==> True
~~~~~

As to why that is specifically needed, I am not certain.

Reply:
That could diverge the behaviour for `count == bil`:
Assuming, `bil = 64, count = 64, `limbs =1`

~~~~~~
v0 = count / (biL);        ==> 1
v1 = count & (biL - 1);    ==> 0

(v0 > limbs || (v0 == limbs && v1 > 0))    ==> False (Because v1 == v0)
(v0 >= limbs)                              ==> True (Because v0 == limbs )
~~~~~~

I believe that is a valid use case?

Reply:
In the long run we would like to make this constant time and maybe even optimise it. But for now, we are happy with the performance. What could be useful is, to be similar to the original `shift_l` as this would reduce the risk of regressions.

Reply:
Admittedly the original `shift_l` had no such checks, so I guess this is good for now. The checks remain in `shift_r`, however - but this is not in scope for this PR.